### DISCUSSION TITLE
gramine-sgx hold at step of loading

### DISCUSSION THREAD

**Discussion by chongloubbt** [gramineproject/gramine#1259] [General]
When I exec a program "xsce" with gramine-sgx, it hold at step of loading…

Is there any log to debug this? And what conditions can cause this problem？

[P1:T1:xsce] debug: Creating pipe: pipe.srv:33ae186edfd95bdcdd6f65630930dcc497bdda64c078b8cf83fd0fec87158b17
[P1:T1:xsce] debug: Allocated stack at 0x8f6cb000 (size = 0x4000000)
[P1:T1:xsce] debug: loading "file:xsce"

**Comment by dimakuv** [comment]
Could you try to use `loader.log_level = "all"` in the Gramine manifest file? I would find it very surprising if Gramine would just hang on loading a file...

**Comment by chongloubbt** [comment]
I have tried using `loader.log_level = "all"` configuration, but there are no more logs.

**Reply by dimakuv** [reply]
Can you show us the command with which you're running?

And the complete log that Gramine prints.

Also, what is the `xsce` program? Is there anything special about it?

**Comment by chongloubbt** [comment]
[logs.txt](https://github.com/gramineproject/gramine/files/11109810/logs.txt)
The command is `"gramine-sgx xsce"`. 

The `logs.txt` file is the full logs.

The `xsce` program is my test code, and it can be runned success on 2023.03.27.  

I think there may be some limited in resources, but there is no logs to help me to debug.

**Comment by chongloubbt** [comment]
There is not anything special in the program. When I remove some code, it can be runned, but there is nothing special about the removed code.

**Comment by chongloubbt** [comment]
In the manifest, I config 64 thread.  gramine-sgx can run 64 threads before, but now it hang at loading. The thread is only one as following:

<img width="550" alt="image" src="https://user-images.githubusercontent.com/64451261/228789816-eb8c305b-29b8-48dd-a3cb-272b6bfd9167.png">


loader.pal_internal_mem_size = "64M"
sys.stack.size = "32M"
sys.brk.max_size = "256M"

sys.insecure__allow_eventfd = true
sys.enable_sigterm_injection = true
sgx.thread_num = 64
sgx.debug = true
sgx.enclave_size = "4G"
sgx.max_threads = 64
sgx.file_check_policy = "allow_all_but_log" 
sgx.nonpie_binary = true

**Comment by dimakuv** [comment]
> it can be runned success on 2023.03.27.

What does this mean? What happened after this date? You modified the `xsce` program?

Looking at your log, you're using v1.3.1 of Gramine (`debug: Gramine was built from commit: 1.3.1`). Could you try the latest Gramine?

Also, have you tried running with `gramine-direct`? Does it work?

**Comment by chongloubbt** [comment]
1. I decrease the threads num in the `xsce` program,  the modification should not affect program. And it has not yet run to the place where I modified it.
2. gramine-direct mode can run success.
3. 1.4 can't run success, because  `is_oot` in `gramine-sgx-get-token` check failed.

**Comment by chongloubbt** [comment]
When I gdb attach the `loader of gramine-sgx`, it seems hang in `async_exit_pointer` as following:

<img width="986" alt="image" src="https://user-images.githubusercontent.com/64451261/228803301-0bcb2871-197c-420f-bcdc-060c8df55dc6.png">

<img width="724" alt="image" src="https://user-images.githubusercontent.com/64451261/228803568-f2fdf24f-5bd9-4884-bc0b-db23a0b2f0bd.png">

**Reply by dimakuv** [reply]
GDB on the release version of Gramine won't show anything interesting. You'll need to build Gramine from sources and in debug mode (see https://gramine.readthedocs.io/en/stable/devel/building.html).

> 1.4 can't run success, because is_oot in gramine-sgx-get-token check failed.

I don't know what you mean by this, but you should try to resolve this issue first. Are you using a very old Intel CPU? Why do you need to OOT legacy SGX driver?

**Reply by chongloubbt** [reply]
`gramine-sgx-get-token` will `return` in the following code.

<img width="593" alt="image" src="https://user-images.githubusercontent.com/64451261/228805586-a3ca221a-39ff-4e9f-8170-d36d272a8635.png">

**Reply by chongloubbt** [reply]
I have tried to build source code follow the building doc, but there are many building errors.

**Comment by chongloubbt** [comment]
`enclave_size` is set as 32G，the example of ra-tls-mbedtls hang on, But `enclave_size` is set as 8G，it can run.  Is there any info to debug that?

**Reply by dimakuv** [reply]
What is your machine? How much RAM does it have (the output of `free -h`)? How much EPC does it have? You can run `is-sgx-available` to learn about the EPC size.

I have a suspicion that you have little EPC, and probably also little RAM, so trying to run a 32GB enclave leads to page swapping and dramatic drop in performance.

**Reply by chongloubbt** [reply]
root@ubuntu:/home# is-sgx-available 
SGX supported by CPU: true
SGX1 (ECREATE, EENTER, ...): true
SGX2 (EAUG, EACCEPT, EMODPR, ...): true
Flexible Launch Control (IA32_SGXPUBKEYHASH{0..3} MSRs): true
SGX extensions for virtualizers (EINCVIRTCHILD, EDECVIRTCHILD, ESETCONTEXT): false
Extensions for concurrent memory management (ETRACKC, ELDBC, ELDUC, ERDINFO): false
CET enclave attributes support (See Table 37-5 in the SDM): false
Key separation and sharing (KSS) support (CONFIGID, CONFIGSVN, ISVEXTPRODID, ISVFAMILYID report fields): true
Max enclave size (32-bit): 0x80000000
Max enclave size (64-bit): 0x100000000000000
EPC size: 0xfe000000
SGX driver loaded: true
AESMD installed: true
SGX PSW/libsgx installed: true


root@ubuntu:/home# free -h
              total        used        free      shared  buff/cache   available
Mem:          247Gi        43Gi        92Gi       717Mi       111Gi       201Gi
Swap:         9.0Gi          0B       9.0Gi

**Reply by dimakuv** [reply]
You have only 4GB of EPC size, which means that any SGX workload above 4GB will be slow. And the greater the size of the workload (like 32GB), the worse the performance will get. This explains why your workload hangs, I guess (have you tried waiting for several hours?).

**Reply by chongloubbt** [reply]
> You have only 4GB of EPC size, which means that any SGX workload above 4GB will be slow. And the greater the size of the workload (like 32GB), the worse the performance will get. This explains why your workload hangs, I guess (have you tried waiting for several hours?).

 On 2023.03.27, it run success after about 15minutes.  Later on 2023.03.28, it hang for more than 10 hours, and I kill it.

**Reply by dimakuv** [reply]
> On 2023.03.27, it run success after about 15minutes. Later on 2023.03.28, it hang for more than 10 hours, and I kill it.

Unfortunately, I do not know what changed in your environment on those days. From what I understand, the Gramine installation was the same. So, it doesn't look like a problem with Gramine.

**Comment by chongloubbt** [comment]
Is there any docker images, and the sgx driver/sdk/psw has been installed in it, then the gramine program can be built direct. Thanks.

**Reply by dimakuv** [reply]
There is one Docker image that we publish -- https://hub.docker.com/r/gramineproject/gramine (currently has Gramine v1.4). Please also read the README carefully on that web page.

**Reply by chongloubbt** [reply]
Thank you very much.  
Can the Docker image be used to build the gramine source code？  Or can gramine source code can be built in docker？When I build the source code in the docker, it depend sgx.h and some more files of sgx driver. There is not those files in the docker.

**Reply by dimakuv** [reply]
You can circumvent the problem of missing `sgx.h` by just downloading it into some directory and specifying it during `meson build`, as is done here: https://github.com/gramineproject/gsc/blob/77b1f70e297a286a82708ec6c76b8e4243a5c1bd/templates/Dockerfile.common.compile.template#L22-L37

Alternatively, you can expose some volumes from the host OS to the Docker container, but I don't have a snippet at hand that you could copy-paste. (But this is a cleaner approach than the first one, if you want to do it properly.)

**Comment by woju** [comment]
> Or can gramine source code can be built in docker？When I build the source
> code in the docker, it depend sgx.h and some more files of sgx driver. There
> is not those files in the docker.

`sgx.h` is from `linux-libc-dev` package, starting from version 5.11. You can
use `debian:bullseye-backports` docker image to build gramine just fine
(probably with intel-sgx repo added).

**Reply by chongloubbt** [reply]
OK, thanks.

**Comment by chongloubbt** [comment]
Run with gramine of debug mode, it seems hang in _PalVirtualMemoryAlloc function? Maybe the size of memset is too large??

<img width="861" alt="image" src="https://user-images.githubusercontent.com/64451261/229566834-5abc7044-293f-4c86-a928-ea6d56e88b0e.png">

**Reply by dimakuv** [reply]
Yeah, as I mentioned previously, it feels like you are limited in EPC, and you have **a lot of page swapping**.

Though one thing I don't understand -- why is it about `xsce`? Can you show `ls -lah` for the `xsce` binary? Could it be that it's so huge?

**Reply by chongloubbt** [reply]
<img width="354" alt="image" src="https://user-images.githubusercontent.com/64451261/229705183-d48564e1-b22d-46ff-b6c2-7c12ac44d5dd.png">

When I build the source code of gramine v1.4 with debug, and it seems hang in `ecall_enclave_start` function.

**Reply by chongloubbt** [reply]
Thans.

I modify the `sgx.enclave_size = "4G"` to `sgx.enclave_size = "2G"`, it can run success.
Maybe this is the reason, can Gramine output some `debug/warn log info` to show this condition.

**Reply by dimakuv** [reply]
> I modify the `sgx.enclave_size = "4G"` to `sgx.enclave_size = "2G"`, it can run success.

So the problem was indeed in the hardware limitation (limited EPC), which resulted in very slow loading of the `xsce` binary? And by decreasing the enclave size, the EPC limitation was not hit? Sorry, I'm just not sure I understand what is really happenning in your case.

> Maybe this is the reason, can Gramine output some `debug/warn log info` to show this condition.

What is the reason? And what would you expect Gramine to output?

**Reply by chongloubbt** [reply]
`So the problem was indeed in the hardware limitation (limited EPC), which resulted in very slow loading of the xsce binary? And by decreasing the enclave size, the EPC limitation was not hit?`   Yes, it is. I decrease the enclave size, it is not hang. 

`What is the reason? And what would you expect Gramine to output?`
—> The EPC size limitation maybe cause gramine hang or slow loading. When this happened, if there some `logs` to show, such as `warning: EPC size may cause program loading to be too slow` or `trace: some function in, some function out`,  the debugging will be faster.
